Approach:
Use the same data as in assignment 1 but this time identify top-10 tokens that occur in regulation descriptions in the table.
1. As in assignment 1, extract regulation descriptions from each record corresponding to a failed inspection
2. Tokenize each regulation description
3. Find top-10 tokens (for the whole table)
4. Clean data: convert to lower case, remove stopwords, punctuation, numbers, etc
5. Find top-10 tokens again
6. Find top-10 tokens after applying Porter stemming to the tokens obtained in step 4.
7. Find top-10 tokens after applying Lancaster stemming to the tokens obtained in step 4.
8. Find top-10 tokens after applying lemmatization to the tokens obtained in step 4.
9. Compare top-10 tokens obtained in 3, 5, 6, 7, 8.

In [52]:
import pandas as pd 
import numpy as np
import re

pd.set_option('display.max_colwidth', None)

## Extract regulation descriptions from each record corresponding to a failed inspection

In [53]:
df = pd.read_csv('Food_Inspections.csv')
df.head()

Inspection ID       DBA Name              AKA Name  License #  \
0        2561123     LOWCOUNTRY            LOWCOUNTRY  1042664.0   
1        2560419     CHARTWELLS  MANSUETO HIGH SCHOOL  2549059.0   
2        2557095        WOW BAO               WOW BAO  1379974.0   
3        2557044  PAN ARTESANAL         PAN ARTESANAL  2602146.0   
4        2556917         BISTRO                BISTRO  2846045.0   

         Facility Type           Risk  \
0           Restaurant  Risk 1 (High)   
1  HIGH SCHOOL KITCHEN  Risk 1 (High)   
2           Restaurant  Risk 1 (High)   
3               Bakery  Risk 1 (High)   
4           Restaurant  Risk 1 (High)   

                                             Address     City State      Zip  \
0                                   3343 N CLARK ST   CHICAGO    IL  60657.0   
1                                  2911 W W 47TH ST   CHICAGO    IL  60632.0   
2                                835 N MICHIGAN AVE   CHICAGO    IL  60611.0   
3                              3724 W FULLERTON AVE   CHICAGO    IL  60647.0   
4  1400 S JEAN BAPTISTE POINTE DUSABLE LAKESHORE DR   CHICAGO    IL  60605.0   

  Inspection Date        Inspection Type          Results Violations  \
0      07/21/2022              Complaint         No Entry        NaN   
1      07/07/2022                Canvass  Out of Business        NaN   
2      06/09/2022  Canvass Re-Inspection             Pass        NaN   
3      06/09/2022                Canvass             Pass        NaN   
4      06/07/2022                Canvass             Pass        NaN   

    Latitude  Longitude                                  Location  
0  41.942869 -87.652863  (41.942869318828365, -87.65286280377227)  
1        NaN        NaN                                       NaN  
2  41.897741 -87.623961  (41.897740856252504, -87.62396131598219)  
3  41.924670 -87.720445   (41.92467025197142, -87.72044496440567)  
4        NaN        NaN                                       NaN

In [54]:
df = df[df['Results'] == 'Fail']
df = df[df['Violations'].notna()]
df = df[['Violations']]

In [55]:
def extract_description(t):
    """
    Extracts description for the violation column
    """
    description_regex = r"\s[A-Z \W]+ -"
    descriptions = re.findall(description_regex, t)
    descriptions = [description[1:-2] for description in descriptions]
    return descriptions

df['descriptions'] = df['Violations'].apply(lambda t: extract_description(t))

In [56]:
df.head()

Violations  \
24                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 6. PROPER EATING, TASTING, DRINKING, OR TOBACCO USE - Comments: 2-401.11  OBSERVED COFFEE CUPS, PIZZA BOXES AND DOUGHNUTS STORED IN THE MEAT PREP AREA ON THE CUTTING BOARDS AND COUNTERTOPS.  MANAGEMENT INSTRUCTED THAT EMPLOYEES MUST EAT AND DRINK IN DESIGNATED AREAS ONLY. | 10. ADEQUATE HANDWASHING SINKS PROPERLY SUPPLIED AND ACCESSIBLE - Comments: 6-301.14  OBSERVED NO HAND WASHING SIGN LOCATED IN THE WASHROOM ADJACENT TO THE MEAT PREP AREA.  MANAGEMENT INSTRUCTED TO INSTALL AND MAINTAIN. | 10. ADEQUATE HANDWASHING SINKS PROPERLY SUPPLIED AND ACCESSIBLE - Comments: 5-204.11  OBSERVED NO EXPOSED HAND WASHING SINK LOCATED AT THE FRONT CASH REGISTER STATION WHERE FETA CHEESE IS SERVED OUT OF THE DELI CASE TO CUSTOMERS. MANAGEMENT INSTRUCTED TO INSTALL AN EXPOSED HAND WASHING SINK IN THIS AREA OR RELOCATE THE COOLER WHERE A HAND WASHING SINK IS CLOSE AND ACCESSIBLE. PRIORITY FOUNDATION 7-38-030(C). NO CITATION ISSUED. | 29. COMPLIANCE WITH VARIANCE/SPECIALIZED PROCESS/HACCP - Comments: 3-502.12 VACUUM PACKAGING DEVICE ON COUNTERTOP IN REAR MEAT PREPARATION AREA.  UNIT TAGGED 'HELD FOR INSPECTION' AT THIS TIME.  MUST OBTAIN CDPH APPROVAL PRIOR TO TAG REMOVAL AND USE OF DEVICE. | 38. INSECTS, RODENTS, & ANIMALS NOT PRESENT - Comments: 6-501.112  OBSERVED DEAD INSECTS ON THE FLOOR OF THE BASEMENT STORAGE AREA.  MANAGEMENT INSTRUCTED TO CLEAN AND REMOVE ALL DEAD INSECTS. | 47. FOOD & NON-FOOD CONTACT SURFACES CLEANABLE, PROPERLY DESIGNED, CONSTRUCTED & USED - Comments: 4-501.12  OBSERVED SOME CUTTING BOARDS IN THE MEAT PREP AND PRODUCE PREP AREA THAT ARE HEAVILY SCORED AND DISCOLORED.  MANAGEMENT INSTRUCTED TO RESURFACE OR REPLACE THE CUTTING BOARDS. | 47. FOOD & NON-FOOD CONTACT SURFACES CLEANABLE, PROPERLY DESIGNED, CONSTRUCTED & USED - Comments: 4-202.16  OBSERVED MILK AND SODA CRATES USED FOR STORAGE IN THE WALK-IN COOLER WITH ACCUMULATED FOOD DEBRIS UNDER THE CRATES.  MANAGEMENT INSTRUCTED TO REMOVE AND REPLACE WITH SOMETHING WHICH ALLOWS FOR EASY CLEANING, SUCH AS RAISED SHELVING UNITS. | 49. NON-FOOD/FOOD CONTACT SURFACES CLEAN - Comments: 4-602.13  CLEAN THE TOASTER OVEN IN THE PRODUCE PREP AREA, AND THE FAN COVERS IN THE MEAT WALK-IN COOLER. | 53. TOILET FACILITIES: PROPERLY CONSTRUCTED, SUPPLIED, & CLEANED - Comments: 6-501.19  THE SELF-CLOSING DOOR DEVICE FOR THE WASHROOM DOOR IS NOT INSTALLED.  MANAGEMENT INSTRUCTED TO REPAIR AND MAINTAIN. | 54. GARBAGE & REFUSE PROPERLY DISPOSED; FACILITIES MAINTAINED - Comments: 5-501.110  OBSERVED USED CUTTING BOARDS AND GARBAGE AND LITTER ON THE GROUND SURROUNDING THE GARBAGE DUMPSTERS.  MANAGEMENT INSTRUCTED TO CLEAN AND MAINTAIN THE OUTSIDE GARBAGE AREA AT ALL TIMES.  PRIORITY FOUNDATION 7-38-020(B).  CITATION ISSUED. | 55. PHYSICAL FACILITIES INSTALLED, MAINTAINED & CLEAN - Comments: 6-501.114  REMOVE ALL UNNECESSARY ITEMS FROM UNDER THE STAIRCASE TO THE OFFICE AND BASEMENT. STORE ALL ELSE TO PREVENT PEST HARBORAGE AND TO FACILITATE CLEANING. | 55. PHYSICAL FACILITIES INSTALLED, MAINTAINED & CLEAN - Comments: 6-501.12  CLEAN THE WALLS IN THE MEAT PREP AREAS AND PRODUCE PREP AREA WITH FOOD DEBRIS AND SPLATTER.  CLEAN THE FLOOR THROUGHOUT THE PREP AREAS, WALK-IN COOLERS AND BASEMENT. REMOVE ALL LITTER AND DEBRIS. | 55. PHYSICAL FACILITIES INSTALLED, MAINTAINED & CLEAN - Comments: 6-501.16  OB

## Tokenize each regulation description

In [57]:
import nltk
import nltk.corpus  
from nltk.tokenize import word_tokenize
from nltk.text import Text

In [58]:
list_ = np.array(df['descriptions'])
tokens = Text(list_)
tokens = np.concatenate(tokens)
# words = nltk.tokenize.word_tokenize(tokens)
tokens = ", ".join(tokens)
tokens[:200]

'PROPER EATING, TASTING, DRINKING, OR TOBACCO USE, ADEQUATE HANDWASHING SINKS PROPERLY SUPPLIED AND ACCESSIBLE, ADEQUATE HANDWASHING SINKS PROPERLY SUPPLIED AND ACCESSIBLE, COMPLIANCE WITH VARIANCE/SPE'

In [59]:
words = word_tokenize(tokens)
words[:15]

['PROPER',
 'EATING',
 ',',
 'TASTING',
 ',',
 'DRINKING',
 ',',
 'OR',
 'TOBACCO',
 'USE',
 ',',
 'ADEQUATE',
 'HANDWASHING',
 'SINKS',
 'PROPERLY']

## Find top-10 tokens (for the whole table)

In [60]:
fdist = nltk.FreqDist(words)
print(fdist)

#fdist.items() - will give all words
token_all = fdist.most_common(10)

<FreqDist with 3025 samples and 3730531 outcomes>


## Clean data: convert to lower case, remove stopwords, punctuation, numbers, etc

In [61]:
stopwords = set(nltk.corpus.stopwords.words('english'))
list(stopwords)[:10]

['hers',
 'if',
 'm',
 "that'll",
 'further',
 'after',
 'in',
 'few',
 'once',
 "mightn't"]

In [66]:
words=[word.lower() for word in words if (word.isalpha()) & (word not in stopwords)]
words[:20]

['proper',
 'eating',
 'tasting',
 'drinking',
 'tobacco',
 'use',
 'adequate',
 'handwashing',
 'sinks',
 'properly',
 'supplied',
 'accessible',
 'adequate',
 'handwashing',
 'sinks',
 'properly',
 'supplied',
 'accessible',
 'compliance',
 'insects']

In [67]:
fdist = nltk.FreqDist(words)
print(fdist)

token_preprocessed = fdist.most_common(10)
token_preprocessed

<FreqDist with 2057 samples and 2273832 outcomes>


[('maintained', 83838),
 ('food', 83568),
 ('properly', 66966),
 ('clean', 66375),
 ('constructed', 65594),
 ('equipment', 64487),
 ('installed', 63103),
 ('cleaning', 48289),
 ('surfaces', 47402),
 ('contact', 44559)]

## Find top-10 tokens after applying Porter stemming to the tokens obtained in step 4.

In [69]:
porter = nltk.PorterStemmer()

token_porter = [porter.stem(t) for t in words]
fdist = nltk.FreqDist(token_porter)

token_porter = fdist.most_common(10)
token_porter

[('clean', 140318),
 ('food', 88048),
 ('maintain', 87945),
 ('properli', 66966),
 ('construct', 65594),
 ('equip', 64493),
 ('instal', 63162),
 ('surfac', 47463),
 ('contact', 44560),
 ('method', 40861)]

## Find top-10 tokens after applying Lancaster stemming to the tokens obtained in step 4.

In [71]:
lancaster = nltk.LancasterStemmer()

token_lancaster = [lancaster.stem(t) for t in words]
fdist = nltk.FreqDist(token_lancaster)

token_lancaster = fdist.most_common(10)
token_lancaster

[('cle', 147846),
 ('food', 88048),
 ('maintain', 87945),
 ('prop', 82874),
 ('construct', 65594),
 ('equip', 64493),
 ('instal', 63162),
 ('surfac', 47463),
 ('contact', 44560),
 ('method', 40861)]

## Find top-10 tokens after applying lemmatization to the tokens obtained in step 4.

In [72]:
wnl = nltk.WordNetLemmatizer()

token_lemma = [wnl.lemmatize(t) for t in words]
fdist = nltk.FreqDist(token_lemma)

token_lemma = fdist.most_common(10)
token_lemma

[('food', 88048),
 ('maintained', 83838),
 ('properly', 66966),
 ('clean', 66375),
 ('constructed', 65594),
 ('equipment', 64491),
 ('installed', 63103),
 ('cleaning', 48289),
 ('surface', 47463),
 ('contact', 44559)]

## Compare top-10 tokens obtained in 3, 5, 6, 7, 8.

In [73]:
token_all

[(',', 644165),
 ('AND', 170713),
 (':', 125580),
 ('MAINTAINED', 83838),
 ('FOOD', 83568),
 ('PROPERLY', 66966),
 ('CLEAN', 66375),
 ('CONSTRUCTED', 65594),
 ('EQUIPMENT', 64487),
 ('&', 63424)]

In [74]:
token_preprocessed

[('maintained', 83838),
 ('food', 83568),
 ('properly', 66966),
 ('clean', 66375),
 ('constructed', 65594),
 ('equipment', 64487),
 ('installed', 63103),
 ('cleaning', 48289),
 ('surfaces', 47402),
 ('contact', 44559)]

In [75]:
token_porter

[('clean', 140318),
 ('food', 88048),
 ('maintain', 87945),
 ('properli', 66966),
 ('construct', 65594),
 ('equip', 64493),
 ('instal', 63162),
 ('surfac', 47463),
 ('contact', 44560),
 ('method', 40861)]

In [76]:
token_lancaster

[('cle', 147846),
 ('food', 88048),
 ('maintain', 87945),
 ('prop', 82874),
 ('construct', 65594),
 ('equip', 64493),
 ('instal', 63162),
 ('surfac', 47463),
 ('contact', 44560),
 ('method', 40861)]

In [77]:
token_lemma

[('food', 88048),
 ('maintained', 83838),
 ('properly', 66966),
 ('clean', 66375),
 ('constructed', 65594),
 ('equipment', 64491),
 ('installed', 63103),
 ('cleaning', 48289),
 ('surface', 47463),
 ('contact', 44559)]

# Discussion

## Unprocessed data is containing the punctuations and stopwords that were most frequently used in the claims. After processing the text, we see distinct words that provide some context about the claims. For example, we observe that the top 10 words are generally positive, yet the word "clean" might have been used as "not clean". Therefore, we lack the sufficient evidence to conclude whether the business are good or bad. Since we know these are the descriptions from failed inspections, we can say that there might be problems with the food and maintenance of equipments.



## Using the Porter's Stemmer increased the occurance of clean. This might be the result of changing ["cleaning", "cleaned"] to "clean".


In [87]:
Text(words).concordance('cleaning')

Displaying 25 of 48289 matches:
eaned good repair coving installed cleaning methods used walls ceilings attach
er code good repair surfaces clean cleaning methods lighting required minimum 
d free litter unnecessary articles cleaning equipment properly stored food han
eaned good repair coving installed cleaning methods used walls ceilings attach
er code good repair surfaces clean cleaning methods lighting required minimum 
d free litter unnecessary articles cleaning equipment properly stored sanitizi
eaned good repair coving installed cleaning methods used walls ceilings attach
er code good repair surfaces clean cleaning methods lighting required minimum 
d free litter unnecessary articles cleaning equipment properly stored food ice
eaned good repair coving installed cleaning methods used walls ceilings attach
er code good repair surfaces clean cleaning methods ventilation rooms equipmen
d free litter unnecessary articles cleaning equipment properly stored faciliti
eaned good repair co

## Here we see that, the word "cleaning" was used 48289, so this result proves our hypothesis.

## For the Lancaster's Stemmer, we don't see any distinction between Porter and Lancaster

## After applying the Lemmatizer, we see that the word "food" becomes the top token in our data. This can be a result of changing "foods" to "food"

In [88]:
Text(words).concordance('foods')

Displaying 25 of 4480 matches:
 labeled original container observed foods coolers stored original containers w
labeling must provide name prep date foods stored original containers utensils 
iner must provide identifying labels foods stored original containers inside co
 disclosure reminder menu raw cooked foods provide maintian priority foundation
d manager site potentially hazardous foods prepared served dish machines provid
d manager site potentially hazardous foods prepared served food original contai
d manager site potentially hazardous foods prepared served food original contai
d manager site potentially hazardous foods prepared served food contact surface
ored source sound condition spoilage foods properly labeled shellfish tags plac
d manager site potentially hazardous foods prepared served clean utensils singl
d manager site potentially hazardous foods prepared served dish washing facilit
d manager site potentially hazardous foods prepared served food original contai
d manager

## Here, we can clearly see there were 4480 matches for "foods" which are changed to "food" after applying WordNetLemmatizer